### ISSUE TITLE
IDE0039 (Use local function) should not be reported, if casting would be required

### ISSUE THREAD

**Issue by Dreamescaper** [dotnet/roslyn#38876]
**Version Used**: 
VS2019 16.4 P1

**Steps to Reproduce**:
```csharp
        using System;

        public class Test
        {
            public void Example()
            {
                Func<int, bool> func = _ => true;

                Do(func);
            }

            public void Do(Func<int, bool> f) { }
            public void Do(Predicate<int> p) { }
        }
```

**Expected/Actual Behavior**:
IDE0039 is suggested, and codefix is available - it introduces local function, and then adds explicit cast to `Func`:
```
static bool func(int _) => true;
Do((Func<int, bool>)func);
```

It probably makes no sense, right?

**Comment by CyrusNajmabadi** [comment]
I'm not getting the problem.  Could you clarify why you think "It probably makes no sense, right?" ?

Thanks!

**Comment by Dreamescaper** [comment]
Code is less readable and uglier, in my opinion.
Code with local function is usually cleaner and shorter over Func variable, but not the case if it requires casting to Func.

**Comment by CyrusNajmabadi** [comment]
> Code is less readable and uglier, in my opinion.

I'm not seeing the less-readable/uglier issue.  But that's ultimately a subjective decision.  

IN this case, the suggestion seems correct and reasonable.  Ultimately, it's up to you if you want to adopt the suggestion or not.  That's why they're suggestions and not warnings/errors/etc.  They're a notification that this is possible and something you can consider.  They're not a statement that you must do this.

**Comment by CyrusNajmabadi** [comment]
This is working as expected.
